Doc/c-api/mapping.rst

```diff
@@ -33,6 +33,32 @@ See also :c:func:`PyObject_GetItem`, :c:func:`PyObject_SetItem` and
    See also :c:func:`PyObject_GetItem`.
 
 
+.. c:function:: int PyMapping_GetOptionalItem(PyObject *obj, PyObject *key, PyObject **result)
+
+   Replacement of :c:func:`PyObject_GetItem` which doesn't raise :exc:`KeyError`.
+
+   Return ``1`` and set ``*result != NULL`` if a key is found.
+   Return ``0`` and set ``*result == NULL`` if a key is not found;
+   a :exc:`KeyError` is silenced.
+   Return ``-1`` and set ``*result == NULL`` if an error other than
```

Comment:
To me, "KeyError is silenced" suggests internal implementation details that the caller doesn't need to care about, and b) may or may not be true (in the exact-dict case, no KeyError is silenced because it isn't set in the first place). IMO "no KeyError is raised" is a clearer statement of the contract that is relevant to the caller.

```suggestion
   Return ``1`` and set ``*result != NULL`` if the key is found.
   Return ``0`` and set ``*result == NULL`` if the key is not found;
   no :exc:`KeyError` is raised.
   Return ``-1`` and set ``*result == NULL`` if an error other than
```

Reply:
On second thought, it may be important to specify "the KeyError is silenced" to clarify the behavior with a custom `__getitem__` method that raises `KeyError` explicitly. I take back that part of the suggestion; I would just replace all three instances of `a` with `the`.

Reply:
I dislike specify that ``*pvalue``  is set to NULL on error. I would prefer to leave it undefined.

Code like ``(void)PyMapping_GetOptionalItem(modules, name, &m);`` would be unsafe: I prefer to handle explicitly the error case and no rely on *m* value.

In the implementation, I prefer to set it to NULL to avoid... the undefined behavior :-) Maybe in debug mode we could even set the value to a marker value to detect the usage of the undefined value.

Reply:
I disagree. It will require explicitly setting `result = NULL` before every call of this function, and if you forgot to do this, it will crash your code in rare case of error, which may be not covered by tests. The risk of writing errorenous code is too high even if there would be a benefit from not setting the result to NULL.

Also, it makes the code simpler. Instead of saving the return code into a variable and checking it in two `if`\ s
```c
int rc = PyMapping_GetOptionalItem(...);
if (rc < 0) {
    goto error;
}
if (rc == 0) {
    // handle default
}
```
you can get rid of that variable and insert the call directly in the `if`.
```c
if (PyMapping_GetOptionalItem(...) < 0) {
    goto error;
}
if (value == NULL) {
    // handle default
}
```
And there are many other cases in which it is convenient that you have two ways to check for result. I suggest you to make `PyMapping_GetOptionalItem()` not setting the result to NULL on error and to look how much code it will require to rewrite and how much uglier it will become. It is the same for `PyObject_GetOptionalAttr`.

Reply:
> I disagree. It will require explicitly setting result = NULL before every call of this function, and if you forgot to do this, it will crash your code in rare case of error, which may be not covered by tests.

+1 and slightly related, I have a PR for the devguide repo regarding explicitly clearing output parameters: https://github.com/python/devguide/pull/1129

cc. @serhiy-storchaka, @carljm